renderer.js

```diff
@@ -1,160 +1,188 @@
-const baudRate = 9600;
+const baudRate = 115200;
 const maxValue = 1024;
-const serialport = require('serialport');
+const serialport = require("serialport");
+const Readline = require("@serialport/parser-readline");
 let activePort;
+let parser;
 let nSteps = -1;
-let stepDivs = [];
-let canvases = [];
-let textValues = [];
-let textThresholds = [];
-let interval;
+
+let readings = [];
+let thresholds = [];
+
+let intervalHandle;
+
+// HTML elements
+let stepDivElements = [];
+let canvasElements = [];
+let textValueElements = [];
+let textThresholdElements = [];
 
 const init = async () => {
-	await serialport.list().then((ports) => {
-		const options = ports.map((port) => {
-			const option = document.createElement('option');
-			option.appendChild(document.createTextNode(port.path));
-			return option;
-		});
-		const selectedIndex = ports.findIndex((port) =>
-			port.pnpId.startsWith('USB')
-		);
-		if (selectedIndex !== -1) {
-			initializePort(ports[selectedIndex].path);
-			options[selectedIndex].selected = 'selected';
-		}
-		const select = document.querySelector('#ports-dropdown');
-		options.forEach((option) => select.add(option));
-		select.addEventListener('change', (event) => {
-			nSteps = -1;
-			initializePort(event.target.value);
-		});
-	});
+  const ports = await serialport.list();
```

Comment:
When you `await()` stuff you don't need to `.then()` it anymore. It's a little confusing but it automatically does that in the background so that you can pretend it's blocking code even though Javascript can't block.